File: sapphire-perps/src/sapphire_perps/risk/kernel.py
"""The risk kernel: the single chokepoint every order passes through before
it can reach a venue.

Philosophy: **fail closed**. The kernel approves an order only when it can
prove the order is within every configured limit. Missing price data, missing
account state, or an unrecognised symbol all result in rejection — never a
silent pass-through.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from ..config import RiskLimits
from ..types import AccountState, Order


@dataclass
class RiskDecision:
    approved: bool
    reasons: list[str] = field(default_factory=list)

    @classmethod
    def approve(cls, note: str = "within limits") -> RiskDecision:
        return cls(approved=True, reasons=[note])

    @classmethod
    def reject(cls, reasons: list[str]) -> RiskDecision:
        return cls(approved=False, reasons=reasons)

    def __bool__(self) -> bool:  # allows `if decision:`
        return self.approved


def _resulting_signed_size(order: Order, account: AccountState) -> float:
    existing = account.position_for(order.symbol)
    current = existing.signed_size if existing else 0.0
    return current + order.signed_size


class RiskKernel:
    def __init__(self, limits: RiskLimits):
        self.limits = limits

    def evaluate(
        self,
        order: Order,
        price: float,
        account: AccountState,
        daily_pnl: float = 0.0,
    ) -> RiskDecision:
        reasons: list[str] = []
        lim = self.limits

        # --- fail-closed preconditions ----------------------------------
        if account is None:
            return RiskDecision.reject(["no account state available"])
        if price is None or price <= 0:
            return RiskDecision.reject([f"no valid mark price for {order.symbol}"])
        if order.size <= 0:
            return RiskDecision.reject(["order size must be positive"])

        # --- circuit breaker: daily loss limit --------------------------
        # Checked first so a tripped breaker blocks even reduce-only churn that
        # isn't explicitly closing risk. Reduce-only orders are still allowed
        # through so positions can be flattened.
        if daily_pnl <= -abs(lim.daily_loss_limit_usd) and not order.reduce_only:
            reasons.append(
                f"daily loss limit hit (pnl={daily_pnl:.2f} "
                f"<= -{lim.daily_loss_limit_usd:.2f}); only reduce-only allowed"
            )

        # --- symbol allowlist -------------------------------------------
        if order.symbol not in lim.allowed_symbols:
            reasons.append(f"symbol {order.symbol!r} not in allowlist")

        # --- leverage ----------------------------------------------------
        if order.leverage is not None and order.leverage > lim.max_leverage:
            reasons.append(
                f"leverage {order.leverage} exceeds max {lim.max_leverage}"
            )

        order_notional = order.size * price

        # --- reduce-only validity ---------------------------------------
        # A reduce_only order is exempt from the increase-only limits below, so
        # we must PROVE it actually reduces. It reduces only when there is an
        # existing position on the OPPOSITE side. With no position, or one on
        # the same side, "reduce_only" would instead open/increase exposure —
        # reject it rather than let it bypass the caps.
        if order.reduce_only:
            existing = account.position_for(order.symbol)
            if existing is None or existing.size == 0:
                reasons.append("reduce_only order but no open position to reduce")
            elif existing.side is order.side:
                reasons.append(
                    f"reduce_only order on same side as existing {order.symbol} "
                    "position would increase exposure, not reduce it"
                )

        # Increase-only checks are skipped for (validated) reduce-only orders,
        # which can only shrink risk.
        if not order.reduce_only:
            if order_notional < lim.min_order_notional_usd:
                reasons.append(
                    f"order notional ${order_notional:,.2f} below min "
                    f"${lim.min_order_notional_usd:,.2f}"
                )
            if order_notional > lim.max_order_notional_usd:
                reasons.append(
                    f"order notional ${order_notional:,.2f} exceeds max "
                    f"${lim.max_order_notional_usd:,.2f}"
                )

            resulting_notional = abs(_resulting_signed_size(order, account)) * price
            if resulting_notional > lim.max_position_notional_usd:
                reasons.append(
                    f"resulting {order.symbol} position ${resulting_notional:,.2f} "
                    f"exceeds max ${lim.max_position_notional_usd:,.2f}"
                )

            projected_total = self._projected_total_notional(order, account, price)
            if projected_total > lim.max_total_notional_usd:
                reasons.append(
                    f"projected total notional ${projected_total:,.2f} exceeds max "
                    f"${lim.max_total_notional_usd:,.2f}"
                )

            if self._opens_new_position(order, account):
                if account.open_position_count >= lim.max_open_positions:
                    reasons.append(
                        f"already at max open positions ({lim.max_open_positions})"
                    )

        if reasons:
            return RiskDecision.reject(reasons)
        return RiskDecision.approve()

    # ------------------------------------------------------------------
    def _opens_new_position(self, order: Order, account: AccountState) -> bool:
        existing = account.position_for(order.symbol)
        return existing is None or existing.size == 0

    def _projected_total_notional(
        self, order: Order, account: AccountState, price: float
    ) -> float:
        total = 0.0
        touched = False
        for pos in account.positions:
            if pos.symbol == order.symbol:
                touched = True
                signed = pos.signed_size + order.signed_size
                total += abs(signed) * price
            else:
                total += pos.notional
        if not touched:
            total += abs(order.signed_size) * price
        return total

File: quant-perps-core/src/risk/kernel.ts
import type { RiskLimits } from "../config.js";
import {
  type AccountState,
  type Order,
  orderSignedSize,
  positionFor,
  posNotional,
  openPositionCount,
} from "../types.js";

// The single chokepoint every order passes through. Fail closed: an order is
// approved only when it can be proven within every configured limit.
export interface RiskDecision {
  approved: boolean;
  reasons: string[];
}
const approve = (note = "within limits"): RiskDecision => ({
  approved: true,
  reasons: [note],
});
const reject = (reasons: string[]): RiskDecision => ({
  approved: false,
  reasons,
});

function resultingSignedSize(order: Order, account: AccountState): number {
  const existing = positionFor(account, order.symbol);
  const current = existing ? sideSigned(existing) : 0;
  return current + orderSignedSize(order);
}
function sideSigned(p: { side: "long" | "short"; size: number }): number {
  return (p.side === "long" ? 1 : -1) * p.size;
}

export class RiskKernel {
  constructor(private readonly limits: RiskLimits) {}

  evaluate(
    order: Order,
    price: number,
    account: AccountState | null,
    dailyPnl = 0,
  ): RiskDecision {
    const reasons: string[] = [];
    const lim = this.limits;

    // fail-closed preconditions
    if (!account) return reject(["no account state available"]);
    if (!(price > 0)) return reject([`no valid mark price for ${order.symbol}`]);
    if (!(order.size > 0)) return reject(["order size must be positive"]);

    // circuit breaker: daily loss limit (reduce-only still allowed to flatten)
    if (dailyPnl <= -Math.abs(lim.dailyLossLimitUsd) && !order.reduceOnly) {
      reasons.push(
        `daily loss limit hit (pnl=${dailyPnl.toFixed(2)} <= ` +
          `-${lim.dailyLossLimitUsd.toFixed(2)}); only reduce-only allowed`,
      );
    }

    // symbol allow/deny lists
    if (lim.blacklistSymbols.includes(order.symbol))
      reasons.push(`symbol ${order.symbol} is blacklisted`);
    if (lim.allowedSymbols.length > 0 && !lim.allowedSymbols.includes(order.symbol))
      reasons.push(`symbol ${order.symbol} not in allowlist`);

    // per-order leverage field (opt-in; account leverage enforced below)
    if (order.leverage !== undefined && order.leverage > lim.maxLeverage)
      reasons.push(`leverage ${order.leverage} exceeds max ${lim.maxLeverage}`);

    const orderNotional = order.size * price;

    // reduce-only validity: must have an opposite-side position to reduce
    if (order.reduceOnly) {
      const existing = positionFor(account, order.symbol);
      if (!existing || existing.size === 0)
        reasons.push("reduce_only order but no open position to reduce");
      else if (existing.side === order.side)
        reasons.push(
          `reduce_only order on same side as existing ${order.symbol} ` +
            "position would increase exposure, not reduce it",
        );
    }

    if (!order.reduceOnly) {
      if (orderNotional < lim.minOrderNotionalUsd)
        reasons.push(
          `order notional ${fmt(orderNotional)} below min ${fmt(lim.minOrderNotionalUsd)}`,
        );
      if (orderNotional > lim.maxOrderNotionalUsd)
        reasons.push(
          `order notional ${fmt(orderNotional)} exceeds max ${fmt(lim.maxOrderNotionalUsd)}`,
        );

      const resultingNotional =
        Math.abs(resultingSignedSize(order, account)) * price;
      if (resultingNotional > lim.maxPositionNotionalUsd)
        reasons.push(
          `resulting ${order.symbol} position ${fmt(resultingNotional)} ` +
            `exceeds max ${fmt(lim.maxPositionNotionalUsd)}`,
        );

      const projectedTotal = this.projectedTotalNotional(order, account, price);
      if (projectedTotal > lim.maxTotalNotionalUsd)
        reasons.push(
          `projected total notional ${fmt(projectedTotal)} exceeds max ${fmt(lim.maxTotalNotionalUsd)}`,
        );

      // account leverage cap from post-trade notional / equity
      if (account.equity <= 0)
        reasons.push(
          `non-positive account equity (${fmt(account.equity)}); cannot size order`,
        );
      else {
        const effLev = projectedTotal / account.equity;
        if (effLev > lim.maxLeverage)
          reasons.push(
            `effective leverage ${effLev.toFixed(2)}x (${fmt(projectedTotal)} / ` +
              `${fmt(account.equity)} equity) exceeds max ${lim.maxLeverage}x`,
          );
      }

      if (this.opensNewPosition(order, account)) {
        if (openPositionCount(account) >= lim.maxOpenPositions)
          reasons.push(`already at max open positions (${lim.maxOpenPositions})`);
      }
    }

    return reasons.length ? reject(reasons) : approve();
  }

  private opensNewPosition(order: Order, account: AccountState): boolean {
    const existing = positionFor(account, order.symbol);
    return !existing || existing.size === 0;
  }

  private projectedTotalNotional(
    order: Order,
    account: AccountState,
    price: number,
  ): number {
    let total = 0;
    let touched = false;
    for (const pos of account.positions) {
      if (pos.symbol === order.symbol) {
        touched = true;
        const signed = sideSigned(pos) + orderSignedSize(order);
        total += Math.abs(signed) * price;
      } else {
        total += posNotional(pos);
      }
    }
    if (!touched) total += Math.abs(orderSignedSize(order)) * price;
    return total;
  }
}

const fmt = (n: number): string =>
  `$${n.toLocaleString("en-US", { maximumFractionDigits: 2 })}`;
